presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractMapBlock.java

```diff
@@ -373,14 +373,17 @@ private void checkReadablePosition(int position)
         // The number of entries of hashTables array as if they are always built. It's used to calculate the retained size.
         private int expectedEntryCount;
 
-        HashTables(Optional<int[]> hashTables, int expectedEntryCount)
+        private int expectedHashTableCount;
+
+        HashTables(Optional<int[]> hashTables, int expectedEntryCount, int expectedHashTableCount)
```

Comment:
Not familiar with this code so bare with me for the questions... 😂 Is `expectedHashTableCount` the number of hashTables in `Optional<int[]> hashTables`? Why do we need to pass this as a separate argument? If we have to, I'd say put this as the second argument rather than the 3rd makes more sense to me. Is `expectedEntryCount` the number of entries in each hash table or total number of entries? Would be nice to make this more readable.

Reply:
> Not familiar with this code so bare with me for the questions... 😂 Is `expectedHashTableCount` the number of hashTables in `Optional<int[]> hashTables`? 

Yes. :D  Each map(ie. one row of MapBlock) correspond to one hash table. 

 > Why do we need to pass this as a separate argument? 

In Aria Presto, we reuse arrays when possible. This makes the offsets and mapIsNull arrays maybe longer than their actual number of elements when the Blocks are created by StreamReaders. In MapBlock, the length of offsets were used to identify how many hash tables we need to build. Therefore we need to tell the block how many elements there really are in these arrays. In af15a75a87 "Add positionCount parameter to createBlockFromKeyValue" we added positionCount for this purpose. But createMapBlockInternal() treats the positionCount as logical position count for this block, and when we create hash tables, the startOffset + positionCount was used as the number of hash tables which was incorrect. So we need to know how many hash tables we need to build. Another way is to add a offsetsLength field in MapBlock, but other blocks (ArrayBlock, RowBlock) didn't do it and this information is only needed for building hashtables, so we added a new field in HashTables class.

 > If we have to, I'd say put this as the second argument rather than the 3rd makes more sense to me. 

Done. Also moved the field up before expectedEntryCount. 

 > Is `expectedEntryCount` the number of entries in each hash table or total number of entries? Would be nice to make this more readable.

It's the total number of entries. I updated the comments as below, how does it look?

```
        // The number of hash tables. Each map row corresponds to one hash table if it's built.
        private int expectedHashTableCount;

        // The total number of entries of all hashTables as if they are always built. It's used to calculate the retained size.
        private int expectedEntryCount;
```